### PR TITLE
enable eslint hooks rules

### LINKED ISSUE #3886: Runtime change of editor causes a crash
#### Do you want to request a _feature_ or report a _bug_?

I am reporting a bug.

#### What's the current behavior?

https://codesandbox.io/s/slate-reproductions-forked-er268?file=/index.tsx

To reproduce the issue click on Add Editor twice. The first editor is recreated when clicking on the button.
Slate: 0.58.0
Browser: This issue is not browser specific and is reproducible at least in Chromium Version 85.0.4183.102 (Official Build) Arch Linux (64-bit) and Firefox 80.0.1 (64-bit) for Arch Linux.
OS: Arch Linux

#### What's the expected behavior?

I expect changing the editor to work and not crash the app since React docs explicitly mention at https://reactjs.org/docs/hooks-reference.html#usememo that applications should work even if the memoized var is cleared:
> **You may rely on useMemo as a performance optimization, not as a semantic guarantee.** In the future, React may choose to “forget” some previously memoized values and recalculate them on next render, e.g. to free memory for offscreen components. Write your code so that it still works without useMemo — and then add it to optimize performance.

The issue seems to be caused by not including [editor] as a dependency in plenty of places.

### PR DESCRIPTION
**Pull request by aciccarello**
**Description**
A clear and concise description of what this pull request solves. (Please do not just link to a long issue thread. Instead include a clear description here or your pull request will likely not be reviewed as quickly.)

**Issue**
Fixes: (link to issue)

**Example**
A GIF or video showing the old and new behaviors after this pull request is merged. Or a code sample showing the usage of a new API. (If you don't include this, your pull request will not be reviewed as quickly, because it's much too hard to figure out exactly what is going wrong, and it makes maintenance much harder.)

**Context**
If your change is non-trivial, please include a description of how the new logic works, and why you decided to solve it the way you did. (This is incredibly helpful so that reviewers don't have to guess your intentions based on the code, and without it your pull request will likely not be reviewed as quickly.)

Closes #3886

**Checks**
- [x] The new code matches the existing patterns and styles.
- [x] The tests pass with `yarn test`.
- [x] The linter passes with `yarn lint`. (Fix errors with `yarn fix`.)
- [x] The relevant examples still work. (Run examples with `yarn start`.)
- [x] You've [added a changeset](https://github.com/atlassian/changesets/blob/master/docs/adding-a-changeset.md) if changing functionality. (Add one with `yarn changeset add`.)

### CODE REVIEW

File: packages/slate-react/src/components/editable.tsx [hunk 1]
```diff
@@ -181,11 +182,15 @@ export const Editable = (props: EditableProps) => {
     }
   }, [autoFocus])
 
+  let androidInputManager: AndroidInputManager | null
```
[the remark this answers is not in the document]
Comment (jason0x43) [reply packages/slate-react/src/components/editable.tsx:185]: Weird, but good enough, I suppose. 🙂 

The intent might be a bit clearer with a ref, and then eslint wouldn't need it in the dependency list.

```tsx
const androidInputManager = useRef<AndroidInputManager>();

const onDOMSelectionChange = useMemo(() => throttle(() => {
  ... 
  // do something with androidInputManager.current
  ...
}), [editor, readonly, state]);

androidInputManager.current = useAndroidInputManager(...);

[the remark this answers is not in the document]
Comment (aciccarello) [reply packages/slate-react/src/components/editable.tsx]: I added a commit to implement this change. Let me know if the intent seems clear.

File: .eslintrc.json [hunk 2]
```diff
@@ -19,7 +20,7 @@
   "settings": {
     "import/extensions": [".js", ".ts", ".jsx", ".tsx"],
     "react": {
-      "version": "detect"
+      "version": "16"
```
Comment (aciccarello) [inline .eslintrc.json:23]: I think yarn was causing detect to not work

File: packages/slate-react/test/index.spec.tsx [hunk 3]
```diff
@@ -22,16 +22,16 @@ describe('slate-react', () => {
...
         let el: ReactTestRenderer
 
         act(() => {
           el = create(
             <Slate editor={editor} value={value} onChange={() => {}}>
               <Editable
-                renderElement={({ element, children }) => {
-                  React.useEffect(() => mounts(element), [])
+                renderElement={({ children }) => {
+                  useEffect(() => mounts(), [])
```
Comment (aciccarello) [inline packages/slate-react/test/index.spec.tsx:34]: This is meant to be testing the mount call count (hence the empty dependency) but we were passing the element which eslint wants included in the dependency array. It's not needed for the test.

File: packages/slate-react/src/hooks/android-input-manager/use-android-input-manager.ts [hunk 4]
```diff
@@ -26,6 +26,9 @@ export function useAndroidInputManager({
   node,
   ...options
 }: UseAndroidInputManagerOptions) {
+  /* eslint-disable react-hooks/rules-of-hooks */
```
Comment (jason0x43) [inline packages/slate-react/src/hooks/android-input-manager/use-android-input-manager.ts:29]: Another option here would be to move the IS_ANDROID check outside of the function, like

```tsx
export const useAndroidInputManager = !IS_ANDROID ? () => null : ({
  node,
  ...options
}: UseAndroidInputManagerOptions) => {
```

File: packages/slate-react/src/hooks/use-slate-selector.tsx [hunk 5]
```diff
@@ -113,16 +113,23 @@ export function useSlateSelector<T>(
  * Create selector context with editor updating on every editor change
  */
 export function getSelectorContext(editor: Editor) {
```
Comment (jason0x43) [inline packages/slate-react/src/hooks/use-slate-selector.tsx:115]: This is only used in one place. It's probably worth renaming it to `useSelectorContext` so we don't have to disable the hooks rule for a bunch of hook calls.

File: site/examples/code-highlighting.tsx [hunk 6]
```diff
@@ -234,20 +237,23 @@ const SetNodeToDecorations = () => {
     )
 
     editor.nodeToDecorations = nodeToDecorations
-  }, [editor.children])
+  }, [editor])
```
Comment (jason0x43) [inline site/examples/code-highlighting.tsx:240]: We should be careful here, because `editor.children` might change even if `editor` doesn't.

Comment (aciccarello) [reply site/examples/code-highlighting.tsx]: Good catch. That was significant here. What's confusing is that this function shouldn't have to know about editor.children. This function is setting up a custom `nodeToDecorations` map to optimize some of the other logic. I think it should be triggered by something more explicit than `editor.children` changing.

I can make an eslint exception for now, but the implementation should be thought through. The simplest change that doesn't require an exception would be to drop useMemo.

**Review by dylans — COMMENTED** [review]
Overall looks good, it looks like @jason0x43's feedback covers everything I would have suggested, so please let me know when you've addressed the other feedback. Thanks!

**Pull request merged**